src/mito2/src/memtable/merge_tree/dedup.rs

```diff
@@ -0,0 +1,240 @@
+
+pub trait DedupSource {
+    /// Returns the shard id of source.
+    fn shard_id(&self) -> ShardId;
+
+    /// Returns whether current source is still valid.
+    fn is_valid(&self) -> bool;
+
+    /// Advances source to next data batch.
+    fn next(&mut self) -> Result<()>;
+
+    /// Returns current pk index.
+    /// # Panics
+    /// If source is not valid.
+    fn current_pk_id(&self) -> PkIndex;
+
+    /// Returns the current primary key bytes.
+    /// # Panics
+    /// If source is not valid.
+    fn current_key(&self) -> &[u8];
+
+    /// Returns the data part.
+    /// # Panics
+    /// If source is not valid.
+    fn current_data_batch(&self) -> DataBatch;
+}
+
+struct DedupReader<T> {
+    prev_batch_last_row: Option<(u16, i64)>,
+    current_batch_range: Option<Range<usize>>,
+    inner: T,
+}
+
+impl<T: DedupSource> DedupReader<T> {
+    fn try_new(inner: T) -> Result<Self> {
+        let mut res = Self {
+            prev_batch_last_row: None,
+            current_batch_range: None,
+            inner,
+        };
+        res.next()?;
+        Ok(res)
+    }
+
+    fn is_valid(&self) -> bool {
+        self.current_batch_range.is_some()
+    }
+
+    /// Returns current encoded primary key.
+    /// # Panics
+    /// If inner reader is exhausted.
+    fn current_key(&self) -> &[u8] {
+        self.inner.current_key()
+    }
+
+    fn current_data_batch(&self) -> DataBatch {
+        let range = self.current_batch_range.as_ref().unwrap();
+        let data_batch = self.inner.current_data_batch();
+        data_batch.slice(range.start, range.len())
+    }
+
+    fn next(&mut self) -> Result<()> {
+        loop {
+            match &mut self.prev_batch_last_row {
+                None => {
+                    // First shot, fill prev_batch_last_row and current_batch_range with first batch.
+                    let current_batch = self.inner.current_data_batch();
+                    let pk_index = current_batch.pk_index();
+                    let (last_ts, _) = current_batch.last_row();
+                    self.prev_batch_last_row = Some((pk_index, last_ts));
+                    self.current_batch_range = Some(0..current_batch.num_rows());
+                    break;
+                }
+                Some(prev_last_row) => {
+                    self.inner.next()?;
+                    if !self.inner.is_valid() {
+                        // Resets current_batch_range if inner reader is exhausted.
+                        self.current_batch_range = None;
+                        break;
+                    }
+                    let current_batch = self.inner.current_data_batch();
+                    let pk_index = current_batch.pk_index();
+                    let (first_ts, _) = current_batch.first_row();
+                    let rows_in_batch = current_batch.num_rows();
+
+                    let (start, end) = if pk_index == prev_last_row.0 && first_ts == prev_last_row.1
```

Comment:
We should compare pk id instead of pk index.

Reply:
> We should compare pk id instead of pk index.

fixed in 85c08249e2